src/empire/object.c

```diff
@@ -82,29 +77,131 @@ void empire_object_load(buffer *buf)
+            }
+        } else {
+            for (int r = RESOURCE_MIN; r < RESOURCE_MAX; r++) {
+                full->city_sells_resource[r] = buffer_read_u8(buf);
+            }
+            for (int r = RESOURCE_MIN; r < RESOURCE_MAX; r++) {
+                full->city_buys_resource[r] = buffer_read_u8(buf);
+            }
         }
         obj->invasion_path_id = buffer_read_u8(buf);
         obj->invasion_years = buffer_read_u8(buf);
-        full->trade40 = buffer_read_u16(buf);
-        full->trade25 = buffer_read_u16(buf);
-        full->trade15 = buffer_read_u16(buf);
+        if (version <= SCENARIO_LAST_UNVERSIONED) {
+            int trade40 = buffer_read_u16(buf);
+            int trade25 = buffer_read_u16(buf);
+            int trade15 = buffer_read_u16(buf);
+            for (int r = RESOURCE_MIN; r < RESOURCE_MAX; r++) {
+                int resource_flag = 1 << r;
+                int amount = 0;
+                if (trade40 & resource_flag) {
+                    amount = 40;
+                } else if (trade25 & resource_flag) {
+                    amount = 25;
+                } else if (trade15 & resource_flag) {
+                    amount = 15;
+                }
+                for (int i = 0; i < 10; i++) {
+                    if (old_sells_resource[i] == r) {
+                        full->city_sells_resource[r] = amount;
+                        break;
+                    }
+                }
+                for (int i = 0; i < 8; i++) {
+                    if (old_buys_resource[i] == r) {
+                        full->city_buys_resource[r] = amount;
+                        break;
+                    }
+                }
+            }
+        }
         buffer_skip(buf, 6);
+
+        if (version > SCENARIO_LAST_UNVERSIONED) {
+            buffer_read_raw(buf, full->city_custom_name, sizeof(full->city_custom_name));
+        }
     }
 
     fix_image_ids();
 }
 
-void empire_object_init_cities(void)
+void empire_object_save(buffer *buf)
+{
+    char *buf_data;
+    if (scenario.empire.id != SCENARIO_CUSTOM_EMPIRE) {
+        buf_data = malloc(sizeof(int));
+        buffer_init(buf, buf_data, sizeof(int));
+        buffer_write_i32(buf, 0);
+        return;
+    }
+    int size_per_obj = 118;
+    int total_size = 0;
+    for (int i = 0; i < MAX_EMPIRE_OBJECTS; i++) {
+        full_empire_object *full = &objects[i];
+        if (full->in_use) {
+            total_size += size_per_obj;
+        } else {
+            total_size += 2;
+        }
+    }
+    buf_data = malloc(total_size);
+    buffer_init(buf, buf_data, total_size);
+
+    for (int i = 0; i < MAX_EMPIRE_OBJECTS; i++) {
+        full_empire_object *full = &objects[i];
+        empire_object *obj = &full->obj;
+        buffer_write_u8(buf, obj->type);
+        buffer_write_u8(buf, full->in_use);
+        if (!full->in_use) {
+            continue;
+        }
+
+        buffer_write_u8(buf, obj->animation_index);
+        buffer_skip(buf, 1);
```

Comment:
Suggestion: since you're changing the version anyway, I think you can get rid of these `buffer_skip`s.